\documentclass[10pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
%\usepackage{setspace} 
%\doublespacing


% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{PLoS2009}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **
%% Please insert a running head of 30 characters or less.  
%% Include it twice, once between each set of braces
\markboth{Evolutionary modeling}{Evolutionary modeling}

%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW
\input{defs.tex}
% no idea why \outside was redefined with a "\right," instead of a normal comma in defs.tex, but am redefining it back here - IH, 5/29/09
\renewcommand{\outside}[7]{\beta_{#1}\left(#2,#3,#4\right)}
%% END MACROS SECTION

\begin{document}

% Title must be 150 words or less
\begin{flushleft}
  {\Large
    \textbf{\titlestring: \supptext{4}}
  }
\\
\authorstring
\end{flushleft}


\newpage
\section{Software}

All tools are available from \darturl\ as part of the \dart\ software package for sequence analysis.

\subsection{Automated grammar construction}
We implemented our model construction algorithm on the three-taxon star phylogeny
in a set of Perl scripts.
Given a singlet transducer modeling ancestral sequences and structures and
a branch transducer modeling structural evolution,
the scripts generate C++ code to create the corresponding jointly-normalized SCFG.
All possible models of structural evolution which can be represented by a Pair SCFG
are permitted as input to the scripts,
allowing for flexible and automated model design.

Given files describing the singlet and branch transducers,
including weights of all transitions which may be functions of evolutionary distance,
the package ComposedTreeTransducer::FourWayComposedTT can automatically generate
the state graph and transition matrix of a multi-sequence model of three extant sequences.
It removes the useless windback $\Snull$ states described in the paper and
introduces effective direct transitions caused by bifurcations with possibly-empty children.
The package ComposedTreeTransducer::TripletSCFG transforms a multi-sequence
model created by ComposedTreeTransducer::FourWayComposedTT into the 
corresponding jointly-normalized three-sequence SCFG
and generates C++ code to build the model.

Example singlet and branch transducers files are provided for a simple 
Pair HMM model, a simple Pair SCFG model and the full TKF Structure Tree model.

\subsection{Reconstruction of ancestral structures}
The program \indiegram\ can perform maximum-likelihood inference on the
three-sequence SCFGs automatically generated by the ComposedTreeTransducer::TripletSCFG package.
Complete or no structural information for the three extant sequences can be supplied
as input.

\subsection{Simulation of RNA family evolution}
The program that we wrote to simulate RNA structural alignments from the TKFST model, \evolsayer, is available as part of the same software distribution as \indiegram\ (the \dart\ package).
Another script (\animateevolsayer) can be used to make animations of the evolving RNA structures, in combination with the RNAplot program in the \viennarna\ package \cite{HofackerEtAl94}.



\bibliography{../latex-inputs/alignment,../latex-inputs/reconstruction,../latex-inputs/duplication,../latex-inputs/genomics,../latex-inputs/ncrna,../latex-inputs/url}



\end{document}
